api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,172 @@
+syntax = "proto3";
+
+import "validate/validate.proto";
+import "envoy/api/v2/core/address.proto";
+
+package envoy.config.rbac.v2alpha;
+
+// Role Based Access Control (RBAC) provides service-level and method-level access control for a service.
+// The RBAC engine authorizes a request by evaluating the request context (expressed in the form of
+// :ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`) against the RBAC policies.
+//
+// RBAC policies are additive. The policies are examined in order. A request is allowed once a matching policy
+// is found (suppose the `action_type` is ALLOW).
+//
+// Here is an example of RBAC configuration. It has two policies:
+// * Service account "cluster.local/ns/default/sa/admin" has full access (empty permission entry means full access)
+//   to the service.
+// * Any user (empty principal entry means any user) can read ("GET") the service at paths with prefix "/products" or
+//   suffix "/reviews" when request header "version" set to either "v1" or "v2".
+//
+//   action_type: ALLOW
+//   policies:
+//     "service-admin":
+//       permissions:
+//       -
+//       principals:
+//         authenticated:
+//           name: "cluster.local/ns/default/sa/admin"
+//     "product-viewer":
+//       permissions:
+//       - paths: [prefix: "/products", suffix: "/reviews"]
+//         methods: ["GET"]
+//         conditions:
+//         - header:
+//             key: "version"
+//             values: [simple: "v1", simple: "v2"]
+//       principals:
+//       -
+//
+message RBAC {
+  // Should we do white-list or black-list style access control.
+  enum ActionType {
+    // The policies grant access to principals. The rest is denied. This is white-list style
+    // access control. This is the default type.
+    ALLOW = 0;
+
+    // The policies deny access to principals. The rest is allowed. This is black-list style
+    // access control.
+    DENY = 1;
+  }
+
+  ActionType action_type = 1;
+
+  // Maps from policy name to policy.
+  map<string, Policy> policies = 2;
+}
+
+// Policy specifies a role and the principals that are assigned/denied the role.
+message Policy {
+ // Required. The set of permissions that define a role.
+ repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+
+ // Required. List of principals that are assigned/denied the role based on “action_type”.
+ repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+message StringMatch {
```

Comment:
At the risk of being too general, I'm wondering if we should have `StringMatch` as a base type in `envoy.types`? I feel this is a pattern that keeps coming up.

Reply:
+1

Reply:
@mattklein123 Any suggestion?

Reply:
Sure go for it.